### PR TITLE
fix(cycle): account for re-entry of all mappables

### PR DESCRIPTION
**Pull request by MickVanDuijn**

### CODE REVIEW

File: src/generator/cycle/cycle.spec.ts [hunk 1]
```diff
@@ -56,3 +56,24 @@ test('take n * |X| X === n * X', () => {
         )
     })
 })
+
+test('|take n cycle(X)| === n', () => {
+    forAll(tuple(natural({ max: 10000 }), mappable(array(unknown(), { minLength: 1 }))), ([n, gen]) => {
+        expect(collect(take(n, cycle(gen)))).toHaveLength(n)
```
Comment (pevisscher) [inline src/generator/cycle/cycle.spec.ts:62]:
```suggestion
        return collect(take(n, cycle(gen))).length === n
```

File: src/generator/cycle/cycle.spec.ts [hunk 2]
```diff
@@ -56,3 +56,24 @@ test('take n * |X| X === n * X', () => {
...
+    forAll(tuple(natural({ min: 1, max: 10 }), mappable(array(unknown(), { maxLength: 0 }))), ([n, gen]) => {
+        expect(() => collect(take(n, cycle(gen, maxEmptyCycles)))).toThrowError(
+            `Exceeded the maximum number of consecutive empty cycles (max: ${maxEmptyCycles})`
+        )
+    })
+})
+
+test('given n === 0, |take n cycle(X)| === 0', () => {
+    forAll(tuple(natural({ max: 0 }), mappable(array(unknown(), { maxLength: 10 }))), ([n, gen]) => {
+        expect(collect(take(n, cycle(gen)))).toHaveLength(0)
```
Comment (pevisscher) [inline src/generator/cycle/cycle.spec.ts:77]:
```suggestion
        return collect(take(n, cycle(gen))).length === 0
```

File: src/generator/cycle/cycle.spec.ts (hunk 1 above)
Comment (pevisscher) [reply src/generator/cycle/cycle.spec.ts:62]: or even `length()` could be used here, but is not necessarily cleaner

File: src/generator/cycle/cycle.ts [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
+    const xss = applicative(toTraversable(xs))
+    let consecutiveEmptyCycles = 0
     while (true) {
-        for (const x of toTraversable(xs)) {
+        let cycleSize = 0
+        for (const x of xss) {
             yield x
+            ++cycleSize
+        }
+        consecutiveEmptyCycles = cycleSize === 0 ? consecutiveEmptyCycles + 1 : 0
```
Comment (pevisscher) [inline src/generator/cycle/cycle.ts:14]:
```suggestion
        consecutiveEmptyCycles += cycleSize === 0 ? 1 : 0
```

File: src/generator/cycle/cycle.ts [hunk 4]
```diff
@@ -1,10 +1,19 @@
+import { applicative } from '../..'
 import type { Mappable, InfiniteGenerator } from '../../type'
 import { toTraversable } from '../../type'
 
-export function* cycle<T>(xs: Mappable<T>): InfiniteGenerator<T> {
+export function* cycle<T>(xs: Mappable<T>, maxConsecutiveEmptyCycles = 10000): InfiniteGenerator<T> {
```
Comment (pevisscher) [inline src/generator/cycle/cycle.ts:5]: why do we want this safety? nothing prevents you from writing an endless while too - but this adds a side effect to the function

File: src/random/types/mappable/mappable.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import type { Mappable } from '../../../type'
+import { toGenerator, toTraversable, toTraverser } from '../../../type'
+import type { Arbitrary, Dependent } from '../../arbitrary'
+import { mapArbitrary } from '../../arbitrary'
+import { oneOf } from '../one-of'
+
+export function mappable<T>(items: Arbitrary<Mappable<T>>): Dependent<Mappable<T>> {
+    return oneOf<Arbitrary<Mappable<T>>[]>(
```
Comment (pevisscher) [inline src/random/types/mappable/mappable.ts:8]: I _think_ we may need to make the items applicative here. Also, I don't think you want to have the `Arbitrary<Mappable<T>>` type, but just a simple `Arbitrary<T>` and compose it with an `array` arbitrary, we're only changing the interface; not the underlying values.

**Review by pevisscher — CHANGES_REQUESTED** [review]
See the comments

File: src/generator/cycle/cycle.ts (hunk 3 above)
Comment (MickVanDuijn) [reply src/generator/cycle/cycle.ts:14]: this is not equivalent: my version is resetting the **consecutive** cycle count once a non-empty cycle has been detected

File: src/iterator/applicative/applicative.ts [hunk 6]
```diff
@@ -1,16 +1,33 @@
...
     // optimization for Array.isArray
-    return 'length' in xs
-        ? (xs as unknown as T[])
-        : {
-              [Symbol.iterator]() {
-                  result ??= collect(xs)
-                  return result[Symbol.iterator]()
-              },
-          }
+    if (isArray(xs)) {
```
[the remark this answers is not in the document]
Comment (MickVanDuijn) [reply src/iterator/applicative/applicative.ts:7]: aye 😁 finding counter-examples has become a bit quicker and more memory efficient

File: src/random/types/mappable/mappable.ts (hunk 5 above)
Comment (MickVanDuijn) [reply src/random/types/mappable/mappable.ts:8]: in usage, you can now pass `array(unknown())` quite easily, which easily gives you full control over the array arbitrary params, but my main reason for implementing this was to make it re-entrant 🙊 so you could potentially have an `Arbitrary<Generator<T>>` and pass that in as well. This allows us to test basically all conversions between the different types by testing `mappable(mappable(array(unknown()))`

File: src/generator/cycle/cycle.ts (hunk 3 above)
Comment (MickVanDuijn) [reply src/generator/cycle/cycle.ts]: (this code is removed in the last commit)

File: src/generator/cycle/cycle.ts (hunk 4 above)
Comment (MickVanDuijn) [reply src/generator/cycle/cycle.ts]: 👍 aaaaand it's gone!

**Review by pevisscher — APPROVED** [review]

**Pull request merged**